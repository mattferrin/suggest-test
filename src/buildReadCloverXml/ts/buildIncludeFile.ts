import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
