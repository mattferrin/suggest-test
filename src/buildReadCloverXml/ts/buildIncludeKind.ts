import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    const isStatementKeeper =
      (file?.statements ?? 0) > 1 && (file?.uncoveredStatements ?? 0) > 0;
    const isConditionalKeeper =
      (file?.conditionals ?? 0) > 1 && (file?.uncoveredConditionals ?? 0) > 0;

    if (!input.flags.includeConditionals) {
      return isStatementKeeper;
    } else if (!input.flags.includeStatements) {
      return isConditionalKeeper;
    } else {
      return isStatementKeeper || isConditionalKeeper;
    }
  };
}
