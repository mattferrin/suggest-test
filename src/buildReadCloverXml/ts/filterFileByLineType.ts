import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
