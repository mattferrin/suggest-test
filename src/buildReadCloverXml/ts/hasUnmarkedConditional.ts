import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function hasUnmarkedConditional(
  input: MeowOutput,
  file: Untrusted<FileArray[number]>
): boolean {
  if (input.flags.includeConditionals) {
    const hasConditional =
      file?.metrics?.some((metric) => {
        return Number.parseInt(metric?.$?.conditionals ?? "0") > 0;
      }) ?? false;

    return (
      hasConditional &&
      (file?.line?.every((line) => {
        return line?.$?.type !== "cond";
      }) ??
        false)
    );
  } else {
    return false;
  }
}
