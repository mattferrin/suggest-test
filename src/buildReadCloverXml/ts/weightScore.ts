import { MeowOutput } from "../../bin/ts/cli";

export function weightScore(
  input: MeowOutput,
  statements: number,
  uncoveredStatements: number,
  conditionals: number,
  uncoveredConditionals: number
): number {
  const weightedStatements =
    statements === 0 ? 0 : Math.pow(uncoveredStatements, 1.3) / statements;
  const weightedConditionals =
    conditionals === 0
      ? 0
      : Math.pow(uncoveredConditionals, 1.3) / conditionals;

  const weightedBalance = weightedStatements + weightedConditionals;

  if (!input.flags.includeConditionals) {
    return weightedStatements;
  } else if (!input.flags.includeStatements) {
    return weightedConditionals;
  } else {
    return weightedBalance;
  }
}
