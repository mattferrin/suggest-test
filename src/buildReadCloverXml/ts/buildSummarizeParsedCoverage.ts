import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
