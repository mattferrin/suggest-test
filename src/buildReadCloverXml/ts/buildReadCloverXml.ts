import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
