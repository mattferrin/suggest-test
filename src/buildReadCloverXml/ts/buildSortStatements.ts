import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
