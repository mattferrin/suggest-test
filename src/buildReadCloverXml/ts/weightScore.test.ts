import { MeowOutput } from "../../bin/ts/cli";
import { weightScore } from "./weightScore";

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: MeowOutput;
  statements: number;
  uncoveredStatements: number;
  conditionals: number;
  uncoveredConditionals: number;
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */

    /** unit under test */
    const result = weightScore(
      input.input,
      input.statements,
      input.uncoveredStatements,
      input.conditionals,
      input.uncoveredConditionals
    );

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "769978f9-3ae3-4966-a706-5f484d05d1d6",
  comment: "returns 0 because 0 statements AND conditionals excluded",
  input: {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  statements: 0,
  uncoveredStatements: "not relevant" as any,
  conditionals: "not relevant" as any,
  uncoveredConditionals: "not relevant" as any,
});
snap({
  stabileSortId: "d154d99f-e253-4da7-9537-788e48182353",
  comment: "returns 0 because 0 conditionals AND statements excluded",
  input: {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  statements: "not relevant" as any,
  uncoveredStatements: "not relevant" as any,
  conditionals: 0,
  uncoveredConditionals: "not relevant" as any,
});
snap({
  stabileSortId: "ab67bf36-4b32-4967-b7cb-3916181f1b68",
  comment: "returns 0 because 0 conditionals AND 0 statements",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  statements: 0 as any,
  uncoveredStatements: "not relevant" as any,
  conditionals: 0,
  uncoveredConditionals: "not relevant" as any,
});

snap({
  stabileSortId: "06842b19-a657-41b6-9214-3d9b94c387e4",
  comment: "returns 1 for 1 uncovered statement WHEN conditionals excluded",
  input: {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  statements: 1,
  uncoveredStatements: 1,
  conditionals: 2,
  uncoveredConditionals: 2,
});
snap({
  stabileSortId: "3b34bd76-56d9-4a28-9163-246a31cf7d95",
  comment: "returns 1 for 1 uncovered conditional WHEN statements excluded",
  input: {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  statements: 2,
  uncoveredStatements: 2,
  conditionals: 1,
  uncoveredConditionals: 1,
});
snap({
  stabileSortId: "e226e9dd-2bce-46f2-8489-d68635fdb4d4",
  comment: "returns 2 for 1 uncovered statement AND 1 uncovered conditional",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  statements: 1,
  uncoveredStatements: 1,
  conditionals: 1,
  uncoveredConditionals: 1,
});

snap({
  stabileSortId: "aa1fb537-79aa-4c10-940c-9c2628786224",
  comment:
    "returns about 0.82 WHEN 2 of 3 statements uncovered AND conditionals excluded",
  input: {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  statements: 3,
  uncoveredStatements: 2,
  conditionals: 9,
  uncoveredConditionals: 9,
});
snap({
  stabileSortId: "7532b751-b93a-4cf4-92ec-c39f771020a1",
  comment:
    "returns about 0.82 WHEN 2 of 3 conditionals uncovered AND statements excluded",
  input: {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  statements: 9,
  uncoveredStatements: 9,
  conditionals: 3,
  uncoveredConditionals: 2,
});
snap({
  stabileSortId: "5868e17d-d3b8-4ab1-a1b0-473084a53758",
  comment:
    "returns about 1.64 WHEN conditionals AND statements are 2 of 3 uncovered",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  statements: 3,
  uncoveredStatements: 2,
  conditionals: 3,
  uncoveredConditionals: 2,
});
