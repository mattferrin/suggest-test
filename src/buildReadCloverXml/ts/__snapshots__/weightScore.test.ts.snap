// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`06842b19-a657-41b6-9214-3d9b94c387e4 1`] = `
{
  "returns 1 for 1 uncovered statement WHEN conditionals excluded": {
    "result": 1,
  },
}
`;

exports[`3b34bd76-56d9-4a28-9163-246a31cf7d95 1`] = `
{
  "returns 1 for 1 uncovered conditional WHEN statements excluded": {
    "result": 1,
  },
}
`;

exports[`5868e17d-d3b8-4ab1-a1b0-473084a53758 1`] = `
{
  "returns about 1.64 WHEN conditionals AND statements are 2 of 3 uncovered": {
    "result": 1.6415258844598883,
  },
}
`;

exports[`7532b751-b93a-4cf4-92ec-c39f771020a1 1`] = `
{
  "returns about 0.82 WHEN 2 of 3 conditionals uncovered AND statements excluded": {
    "result": 0.8207629422299442,
  },
}
`;

exports[`769978f9-3ae3-4966-a706-5f484d05d1d6 1`] = `
{
  "returns 0 because 0 statements AND conditionals excluded": {
    "result": 0,
  },
}
`;

exports[`aa1fb537-79aa-4c10-940c-9c2628786224 1`] = `
{
  "returns about 0.82 WHEN 2 of 3 statements uncovered AND conditionals excluded": {
    "result": 0.8207629422299442,
  },
}
`;

exports[`ab67bf36-4b32-4967-b7cb-3916181f1b68 1`] = `
{
  "returns 0 because 0 conditionals AND 0 statements": {
    "result": 0,
  },
}
`;

exports[`d154d99f-e253-4da7-9537-788e48182353 1`] = `
{
  "returns 0 because 0 conditionals AND statements excluded": {
    "result": 0,
  },
}
`;

exports[`e226e9dd-2bce-46f2-8489-d68635fdb4d4 1`] = `
{
  "returns 2 for 1 uncovered statement AND 1 uncovered conditional": {
    "result": 2,
  },
}
`;
